src/libcore/slice/mod.rs

```diff
@@ -1618,6 +1618,45 @@ impl<T> [T] {
         }
     }
 
+    /// Copies elements from one part of a slice to another part of the same
+    /// slice, using a memmove.
+    ///
+    /// `src` is the starting index of the source region. `dest` is the starting
+    /// index of the destination region. `count` is the number of elements in
+    /// both regions. The two regions may overlap.
+    ///
+    /// The ends of the two regions, `src + count` and `dest + count`, must be
+    /// less than or equal to `self.len()`.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if either region exceeds the end of the slice.
+    ///
+    /// # Examples
+    ///
+    /// Copying four bytes within a slice:
+    ///
+    /// ```
+    /// # #![feature(copy_in_place)]
+    /// let mut bytes = *b"Hello, World!";
+    ///
+    /// bytes.copy_in_place(1, 8, 4);
+    ///
+    /// assert_eq!(&bytes, b"Hello, Wello!");
+    /// ```
+    #[unstable(feature = "copy_in_place", issue = "54236")]
+    pub fn copy_in_place(&mut self, src: usize, dest: usize, count: usize) where T: Copy {
+        // Avoid computing `src + count` or `dest + count`, which could overflow.
+        assert!(count <= self.len());
+        assert!(src <= self.len() - count);
+        assert!(dest <= self.len() - count);
+        unsafe {
+            let src_ptr = self.get_unchecked(src) as *const T;
+            let dest_ptr = self.get_unchecked_mut(dest) as *mut T;
```

Comment:
nit: You shouldn't need to cast these; they'll coerce just fine.

Reply:
I think without the casts this fails the borrow checker.

Reply:
Ah, yes, you're of course right.  I was thinking of [when they're not in variables](https://play.rust-lang.org/?gist=417fc09eb4dea89b518c1cc60f00adbf&version=stable&mode=debug&edition=2015):
```rust
        ptr::copy(
            slice.get_unchecked(src),
            slice.get_unchecked_mut(dest),
            count,
        );
```